pkg/server/rbac/service_accounts.go

```diff
@@ -279,41 +279,56 @@ func (s *serviceAccountsDatabase) CreateToken(
 		},
 		Type: corev1.SecretTypeServiceAccountToken,
 	}
-	if err := s.client.Create(ctx, tokenSecret); err != nil {
+	if err = s.client.Create(ctx, tokenSecret); err != nil {
 		return nil, fmt.Errorf(
 			"error creating token Secret %q for ServiceAccount %q in namespace %q: %w",
 			tokenName, saName, namespace, err,
 		)
 	}
-	// Retrieve Secret -- this is necessary to actually get the token. We use
-	// a simple retry because token data is created asynchronously and we don't
+
+	// Retrieve Secret -- this is necessary to actually get the token. We wrap
+	// this in a retry because token data is created asynchronously and we don't
 	// want to prematurely return the Secret without its data.
-	for range 5 {
-		tokenSecret = &corev1.Secret{}
-		if err := s.client.Get(
-			ctx,
-			client.ObjectKey{
-				Namespace: namespace,
-				Name:      tokenName,
-			},
-			tokenSecret,
-		); err != nil {
-			return nil, fmt.Errorf(
-				"error getting token Secret %q for ServiceAccount %q in namespace %q: %w",
-				tokenName, saName, namespace, err,
-			)
-		}
-		if _, ok := tokenSecret.Data["token"]; ok {
-			return tokenSecret, nil
-		}
-		time.Sleep(time.Second)
+	var gotToken bool
+	backoff := retry.DefaultBackoff
+	backoff.Steps = 5
+	if err = retry.OnError(
+		backoff,
+		func(innerErr error) bool {
+			// i.e. Stop retrying if there was an error or we got the token.
+			return innerErr == nil && !gotToken
```

Comment:
Shouldn't we continue retrying if we receive an error that indicates the server is, for example, temporarily unavailable?

Reply:
We could, but my thought process here was that I wasn't out to retry on error. (In the vast majority of places, we don't, after all.) The only thing I was after here was accounting for the (highly probable, as it turns out) scenario where the token hasn't actually been generated yet. But if you feel strongly, I can re-evaluate that.

Reply:
Fixed in 847547e79cf83f5e57e5818fea8471ab15652981